File: java/binary-search/src/main/java/BinarySearch.java
import java.util.ArrayList;
import java.util.List;

class BinarySearch {

    private List<Integer> items;

    BinarySearch(List<Integer> items) {
        this.items = new ArrayList<>(items);
    }

    int indexOf(int item) throws ValueNotFoundException {

        int start = 0;
        int end = this.items.size() - 1;

        while (start <= end) {
            int middle = (start + end) / 2;
            int middleItem = this.items.get(middle);

            if (middleItem == item) {
                return middle;
            } 

            if (middleItem > item) {
                end = middle - 1;
            } else {
                start = middle + 1;
            }

        }

        throw new ValueNotFoundException("Value not in array");

    }
}

File: elixir/binary-search/lib/binary_search.ex
defmodule BinarySearch do
  @doc """
    Searches for a key in the tuple using the binary search algorithm.
    It returns :not_found if the key is not in the tuple.
    Otherwise returns {:ok, index}.

    ## Examples

      iex> BinarySearch.search({}, 2)
      :not_found

      iex> BinarySearch.search({1, 3, 5}, 2)
      :not_found

      iex> BinarySearch.search({1, 3, 5}, 5)
      {:ok, 2}

  """

  @spec search(tuple, integer) :: {:ok, integer} | :not_found
  def search({}, _key), do: :not_found
  def search({key}, key), do: {:ok, 0}
  def search({item}, key) when item != key, do: :not_found
  def search(numbers, key) do
   search(numbers, 0, tuple_size(numbers) - 1, key)
  end

  defp search(numbers, start_index, end_index, key) do
    middle = div(start_index + end_index, 2)
    item = elem(numbers, middle)

    cond do
      start_index > end_index -> :not_found
      item === key -> {:ok, middle}
      key > item -> search(numbers, middle + 1, end_index, key)
      key < item -> search(numbers, start_index, middle - 1, key)
    end
  end
end
